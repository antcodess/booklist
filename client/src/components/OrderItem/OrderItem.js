import React from 'react';
import { withRouter } from 'react-router-dom';

import './OrderItem.scss';

const OrderItem = ({ item, history }) => {
  const { routeName, imageUrl, title, quantity, price } = item;

  return (
    <div className='order-item'> 
      <div className='image-container' onClick={() => history.push(`/${routeName}`)}>
        <img src={imageUrl} alt='book' /> 
      </div>
      <span className='title' onClick={() => history.push(`/${routeName}`)}>{title}</span>
      <span className='quantity'>{quantity}</span>
      <span className='price'>${price}</span>
    </div>
  )
};

export default withRouter(OrderItem);