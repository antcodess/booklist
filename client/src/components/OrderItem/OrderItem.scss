.order-item {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-gap: 10px;
  margin-top: 40px;
  text-align: center;

  .image-container {
    cursor: pointer;

    img {
      width: 100%;
      height: auto;
    }
  }
  .title,
  .quantity,
  .price,
  .date {
    margin: auto 0;
  }

  .title {
    cursor: pointer;
  }
}