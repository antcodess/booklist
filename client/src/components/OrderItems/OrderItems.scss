.order-items {
  width: 100%;

  .order-header {
    display: grid;
    grid-template-columns: 1fr 1fr 1fr 1fr 1fr;
    grid-gap: 10px;
    padding-bottom: 20px;
    border-bottom: 2px solid #ededed;
    text-align: center;
    
    .header-block {
      span {
        font-family: "Hurme Geometric Sans 3 W03 Bold";
      }
    }
    
  }
  .total {
    border-top: 2px solid #ededed;
    display: flex;
    justify-content: flex-end;
    padding-top: 10px;
    margin-top: 40px;
    .total-bold {
      font-family: "Hurme Geometric Sans 3 W03 Bold";
    }
  }
}