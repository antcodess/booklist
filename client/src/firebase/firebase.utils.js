import firebase from 'firebase/app';
import 'firebase/firestore';
import 'firebase/auth';

const config = {
  apiKey: "AIzaSyAsGhn_jcLJKl3iYoapykz-RlO7HCtb6lw",
  authDomain: "booklist-a5340.firebaseapp.com",
  databaseURL: "https://booklist-a5340.firebaseio.com",
  projectId: "booklist-a5340",
  storageBucket: "booklist-a5340.appspot.com",
  messagingSenderId: "281234462009",
  appId: "1:281234462009:web:f0e3201b2d8b9027fb84e8",
  measurementId: "G-QG6HJXT43D"
};

firebase.initializeApp(config);

export const auth = firebase.auth();
export const firestore = firebase.firestore();

export const googleProvider = new firebase.auth.GoogleAuthProvider();
googleProvider.setCustomParameters({ prompt: 'select_account' });
export const signInWithGoogle = () => auth.signInWithPopup(googleProvider);

export const addCollectionAndDocuments = async (collectionKey, objectsToAdd) => {
  const collectionRef = firestore.collection(collectionKey);
  
  const batch = firestore.batch();
  objectsToAdd.forEach(obj => {
    const newDocRef = collectionRef.doc()
    batch.set(newDocRef, obj);
  })
  return await batch.commit();
};

export const addCollectionAndDocumentsFromMap = async (collectionKey, objectToAdd) => {
  const collectionRef = firestore.collection(collectionKey);

  const batch = firestore.batch();
  Object.keys(objectToAdd).forEach(key => {
    const newDocRef = collectionRef.doc()
    batch.set(newDocRef, { [key]: objectToAdd[key] })
  });
  return await batch.commit();
};

export const convertProductsSnapshot = products => {
  const transformedCollection = products.docs.map(doc => {
    return {
      routeName: 'products/' + encodeURIComponent(doc.data().title
                                                            .replace(/[\s]/g, '-')
                                                            .replace(/[,*]/g, '')
                                                            .toLowerCase()),
      id: doc.id,
      ...doc.data()
    }
  });
  return transformedCollection;
};

export const convertCategoriesSnapshotToMap = categories => {
  const transformedCollection = categories.docs.map(doc => doc.data());
  return transformedCollection.reduce((acc, collection) => {
    Object.keys(collection).forEach(key => {
      acc[key] = collection[key];
    })
    return acc;
  },{});
};

export const createUserProfileDocument = async (userAuth, additionalData) => {
  if (!userAuth) return;

  const userRef = firestore.doc(`users/${userAuth.uid}`);
  const snapShot = await userRef.get();

  if (!snapShot.exists) {
    const {displayName, email } = userAuth;
    const createdAt = new Date();

    try {
      await userRef.set({
        displayName,
        email,
        createdAt,
        ...additionalData
      })
    } catch (error) {
      console.log('Error creating user', error.message);
    }
  }
  return userRef;
};

export const updateUserProfileDocument = async (userAuth, updateData) => {
  if (!userAuth) return;

  const userRef = firestore.doc(`users/${userAuth.uid}`);
  const snapShot = await userRef.get();
  if(snapShot.exists) {
    try {
      if ('email' in updateData) {
        try {
          const user = auth.currentUser;
          const credential = firebase.auth.EmailAuthProvider.credential(
            userAuth.email, 
            updateData.password
          );
          // Now you can use that to reauthenticate
          await user.reauthenticateWithCredential(credential);
          await user.updateEmail(updateData['email']);
        } catch (error) {
          console.log('Error updating user', error.message);
          return null;
        }
      
      }
      await userRef.update({
        ...updateData
      });
    } catch (error) {
      console.log('Error updating user', error.message);
    }
  }
  return userRef;
};


export const getCurrentUser = () => {
  return new Promise((resolve, reject) => {
    const unsubscribe = auth.onAuthStateChanged(userAuth => {
      unsubscribe();
      resolve(userAuth);
    }, reject);
  })
};

export default firebase;