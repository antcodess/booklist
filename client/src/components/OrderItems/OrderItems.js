import React from 'react';

import OrderItem from '../OrderItem/OrderItem';
import './OrderItems.scss';

export const OrderItems = ({ order }) => {
  const { id, total, products, createdAt } = order;
  const dateObject = new Date(createdAt.toDate());
  const month = dateObject.getMonth() + 1;
  const day = dateObject.getDate();
  const year = dateObject.getFullYear();
  const date = `${month}/${day}/${year}`;

  return (
    <div className='order-items'>
      <div className='order-details'>
        <span className='order-number'>
          <span className='bold'>Order #: </span>
          {id}
        </span>
        <span className='order-date'>
          <span className='bold'>Order Placed: </span>
          {date}
        </span>
      </div>
      <div className='order-header'>
        <div className='header-block'>
          <span>Product</span>
        </div>
        <div className='header-block'>
          <span>Title</span>
        </div>
        <div className='header-block'>
          <span>Quantity</span>
        </div>
        <div className='header-block'>
          <span>Price</span>
        </div>
      </div>
      { products.map(item => <OrderItem key={item.id} item={item} />)}
      <div className='total'>
        <span className='total-bold'>Total: <span>${total}</span></span>
      </div>
    </div>
  );
};

export default OrderItems;