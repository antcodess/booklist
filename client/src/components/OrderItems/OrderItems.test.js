import React from 'react';
import { shallow } from 'enzyme';
import { OrderItems } from './OrderItems';
import OrderItem from '../OrderItem/OrderItem';
import firebase from '../../firebase/firebase.utils';

describe('OrderItems Component', () => {
  let wrapper;
  let mockOrder;
  let mockProducts;
  let dummyFunction = jest.fn();
  
  beforeEach(() => {
    mockProducts = [{id: 1}, {id: 2}, {id: 3}];
    mockOrder = {
      id: 1, 
      total: 100,
      products: mockProducts,
      createdAt: firebase.firestore.Timestamp.now()
    };

    const mockProps = {
      order: mockOrder
    };
    
    wrapper = shallow(<OrderItems { ...mockProps } />);
  });

  it('should render OrderItems Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render same amount of OrderItem components as products', () => {
    expect(wrapper.find(OrderItem).length).toEqual(mockProducts.length);
  });
});
