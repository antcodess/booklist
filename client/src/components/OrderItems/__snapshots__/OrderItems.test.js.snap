// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderItems Component should render OrderItems Component 1`] = `ShallowWrapper {}`;
